src/connector/src/sink/file_sink/mod.rs

```diff
@@ -0,0 +1,230 @@
+///
+/// The choice of writer used during the actual writing process depends on the encode type of the sink.
+enum FileWriterEnum {
+    ParquetFileWriter(AsyncArrowWriter<OpendalWriter>),
+    FileWriter(OpendalWriter),
+}
+
+#[async_trait]
+impl SinkWriter for OpenDalSinkWriter {
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        // Note: epoch is used to name the output files.
+        // Todo: after enabling sink decouple, use the new naming convention.
+        let epoch = self.epoch.ok_or_else(|| {
+            SinkError::Opendal("epoch has not been initialize, call `begin_epoch`".to_string())
+        })?;
+        if self.sink_writer.is_none() {
+            self.create_sink_writer(epoch).await?;
+        }
+        if self.is_append_only {
+            self.append_only(chunk).await
+        } else {
+            // currently file sink only supports append only mode.
+            unimplemented!()
+        }
+    }
+
+    async fn begin_epoch(&mut self, epoch: u64) -> Result<()> {
+        self.epoch = Some(epoch);
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    /// For the file sink, currently, the sink decoupling feature is not enabled.
+    /// When a checkpoint arrives, the force commit is performed to write the data to the file.
+    /// In the future if flush and checkpoint is decoupled, we should enable sink decouple accordingly.
+    async fn barrier(&mut self, is_checkpoint: bool) -> Result<()> {
+        if is_checkpoint && let Some(sink_writer) = self.sink_writer.take() {
+            match sink_writer {
+                FileWriterEnum::ParquetFileWriter(w) => {
+                    let _ = w.close().await?;
+                }
+                FileWriterEnum::FileWriter(mut w) => w.close().await?,
+            };
+        }
+
+        Ok(())
+    }
+
+    async fn update_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl OpenDalSinkWriter {
+    pub fn new(
+        operator: Operator,
+        write_path: &str,
+        rw_schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+        executor_id: u64,
+        encode_type: SinkEncode,
+    ) -> Result<Self> {
+        let arrow_schema = convert_rw_schema_to_arrow_schema(rw_schema)?;
+        Ok(Self {
+            schema: Arc::new(arrow_schema),
+            write_path: write_path.to_string(),
+            pk_indices,
+            operator,
+            sink_writer: None,
+            is_append_only,
+            epoch: None,
+            executor_id,
+            encode_type,
+        })
+    }
+
+    async fn create_object_writer(&mut self, epoch: u64) -> Result<OpendalWriter> {
+        // Todo: specify more file suffixes based on encode_type.
+        let suffix = match self.encode_type {
+            SinkEncode::Json => "json",
+            SinkEncode::Parquet => "parquet",
+            _ => unimplemented!(),
```

Comment:
Is this arm reachable? We should take extra care on all the `unimplemented!()` and `unreachable!()` introduced in this PR because if they turn out to be reachable, it will panic CN.

Reply:
Yes, it's `unreachable!()`, if the ENCODE is not these two(will remove json in this pr), an error will be reported directly on the frontend:
```
Caused by:
  Bind error: connector s3 does not support format Plain with encode Csv
```